quinn-udp/src/windows.rs

```diff
@@ -111,20 +219,92 @@ impl UdpSocketState {
         bufs: &mut [IoSliceMut<'_>],
         meta: &mut [RecvMeta],
     ) -> io::Result<usize> {
-        // Safety: both `IoSliceMut` and `MaybeUninitSlice` promise to have the
-        // same layout, that of `iovec`/`WSABUF`. Furthermore `recv_vectored`
-        // promises to not write unitialised bytes to the `bufs` and pass it
-        // directly to the `recvmsg` system call, so this is safe.
-        let bufs = unsafe {
-            &mut *(bufs as *mut [IoSliceMut<'_>] as *mut [socket2::MaybeUninitSlice<'_>])
+        let wsa_recvmsg_ptr = WSARECVMSG_PTR.expect("valid function pointer for WSARecvMsg");
+
+        // we cannot use [`socket2::MsgHdrMut`] as we do not have access to inner field which holds the WSAMSG
+        let mut ctrl_buf = cmsg::Aligned([0; CMSG_LEN]);
+        let mut source: WinSock::SOCKADDR_INET = unsafe { mem::zeroed() };
+        let mut data = WinSock::WSABUF {
+            buf: bufs[0].as_mut_ptr(),
+            len: bufs[0].len() as _,
         };
-        let (len, _flags, addr) = socket.0.recv_from_vectored(bufs)?;
+
+        let ctrl = WinSock::WSABUF {
+            buf: ctrl_buf.0.as_mut_ptr(),
+            len: ctrl_buf.0.len() as _,
+        };
+
+        let mut wsa_msg = WinSock::WSAMSG {
+            name: &mut source as *mut _ as *mut _,
+            namelen: mem::size_of_val(&source) as _,
+            lpBuffers: &mut data,
+            Control: ctrl,
+            dwBufferCount: 1,
+            dwFlags: 0,
+        };
+
+        let mut len = 0;
+        unsafe {
+            let rc = (wsa_recvmsg_ptr)(
+                socket.0.as_raw_socket() as usize,
+                &mut wsa_msg,
+                &mut len,
+                ptr::null_mut(),
+                None,
+            );
+            if rc == -1 {
+                return Err(io::Error::last_os_error());
+            }
+        }
+
+        let addr = unsafe {
+            let (_, addr) = socket2::SockAddr::try_init(|addr_storage, len| {
+                *len = mem::size_of_val(&source) as _;
+                ptr::copy_nonoverlapping(&source, addr_storage as _, 1);
+                Ok(())
+            })?;
```

Comment:
nit: if we defined `source` as a `sockaddr_storage`, we could use the simpler `SockAddr::new` instead here.

Reply:
This is impossible because socket2 is still on an older version of windows-sys.